mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala

```diff
@@ -0,0 +1,238 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.regression
+
+import java.io.Serializable
+import java.util.Arrays.binarySearch
+
+import org.apache.spark.api.java.{JavaDoubleRDD, JavaRDD}
+import org.apache.spark.rdd.RDD
+
+/**
+ * Regression model for Isotonic regression
+ *
+ * @param features Array of features.
```

Comment:
Need to be more specific about `features` and `labels`. I would rename `features` to `boundaries` and mention that this is monotonic, and rename `labels` to `predictions` because this is not the original labels. The solution to an isotonic regression problem is piecewise linear. The model only needs to store the boundaries and the computed predictions. We can use linear interpolation for values fall between boundaries. For example, if

~~~
boundaries = [1, 2, 4, 5]
predictions = [1.0, 3.0, 3.0, 4.0]
~~~

then

~~~
predict(1.5) == 2.0
predict(3.5) == 3.0
~~~

We should also document the behavior on the semi-open segments, e.g., `predict(-10) == ?`. I suggest using the smallest prediction here, i.e., `predict(-10) = 1.0`.